### PR TITLE
fix: saving listing with uploaded lottery

### PR DESCRIPTION
**Pull request by YazeedLoonat**
This PR addresses #4242 

- [x] Addresses the issue in full

## Description
This fixes a bug where updating a listing with lottery results uploaded was not working

## How Can This Be Tested/Reviewed?
Create or edit a listing such that: 
- it is closed
- it is set to lottery
- the lottery not run through the partner site (you will manually upload lottery results)
- manually uploaded lottery results

go back in to edit the listing -> attempt to save -> it should now save

## Author Checklist:

- [x] Added QA notes to the issue with applicable URLs
- [x] Reviewed in a desktop view
- [ ] Reviewed in a mobile view
- [ ] Reviewed considering accessibility
- [x] Added tests covering the changes
- [ ] Made corresponding changes to the documentation
- [ ] Ran `yarn generate:client` and/or created a migration when required

## Review Process:

- Read and understand the issue
- Ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Either (1) explicitly ask a clarifying question, (2) request changes, or (3) approve the PR, even if there are very small remaining changes, if you don't need to re-review after the updates

### CODE REVIEW

File: api/test/unit/services/listing.service.spec.ts [hunk 1]
```diff
@@ -2391,6 +2399,15 @@ describe('Testing listing service', () => {
         id: 'example id',
         name: 'example name',
       });
+      prisma.listingEvents.findMany = jest.fn().mockResolvedValue([]);
+      prisma.listingEvents.update = jest.fn().mockResolvedValue({
+        id: 'example id',
+        name: 'example name',
+      });
+      prisma.assets.delete = jest.fn().mockResolvedValue({
```
Comment (emilyjablonski) [inline api/test/unit/services/listing.service.spec.ts:2407]: Could we add a test to this PR? Maybe just expecting this to be called?

File: api/src/services/listing.service.ts [hunk 2]
```diff
@@ -1062,6 +1058,47 @@ export class ListingService implements OnModuleInit {
...
+    await Promise.all(
+      assetIds.map(async (assetData) => {
+        if (assetData.fileId) {
+          await this.prisma.assets.delete({
+            where: {
+              id: assetData.fileId,
+            },
+          });
+        }
+      }),
```
Comment (ludtkemorgan) [inline api/src/services/listing.service.ts:1098]: Could this be done as a `deleteMany` where id is in assetId list?

File: api/src/services/listing.service.ts [hunk 3]
```diff
@@ -1207,22 +1246,25 @@ export class ListingService implements OnModuleInit {
...
-                  assets: event.assets
-                    ? {
-                        create: {
-                          ...event.assets,
-                        },
-                      }
-                    : undefined,
-                })),
+                create: dto.listingEvents.map((event) => {
+                  return {
```
Comment (ludtkemorgan) [inline api/src/services/listing.service.ts:1250]: question: Was the syntax without the return causing problems? Or is this just a preferred format?

File: api/src/services/listing.service.ts [hunk 4]
```diff
@@ -1134,6 +1171,8 @@ export class ListingService implements OnModuleInit {
       dto,
       'listingsBuildingAddress',
     );
+    // Delete all assets tied to listing events before creating new ones
+    await this.updateListingEvents(dto.id);
```
Comment (ludtkemorgan) [inline api/src/services/listing.service.ts:1175]: This is the way we are handling other sub tables like addresses and paper applications, so I am cool with just deleting and recreating every time. But did you look into only creating if it doesn't exist in the db? Similar to how we do listingsBuildingSelectionCriteriaFile? The logic can get pretty gnarly that way though

**Review by ludtkemorgan — COMMENTED** [review]
Mostly just questions

File: api/test/unit/services/listing.service.spec.ts (hunk 1 above)
Comment (YazeedLoonat) [reply api/test/unit/services/listing.service.spec.ts:2407]: I ended up writing a separate set of tests for this function, lemme know if that is okay or if you would prefer the tests being part of this

File: api/src/services/listing.service.ts (hunk 2 above)
Comment (YazeedLoonat) [reply api/src/services/listing.service.ts]: nice call out! updated!

File: api/src/services/listing.service.ts (hunk 3 above)
Comment (YazeedLoonat) [reply api/src/services/listing.service.ts]: nah I was trying to figure out what was going on with some console logs and didn't put things back properly

File: api/src/services/listing.service.ts (hunk 4 above)
Comment (YazeedLoonat) [reply api/src/services/listing.service.ts:1179]: yeah I think we should address that separately, this is something I think I mentioned but with prisma extensions ( https://www.prisma.io/docs/orm/prisma-client/client-extensions )

we could write a custom query that handles that for us, so we can sort simplify that into helper functions basically 

but I think we should do that in a separate ticketo with a spike to test out extensions a bit

**Review by emilyjablonski — APPROVED** [review]
➰

**Pull request merged**